docs/refman/generatorman.py

```diff
@@ -0,0 +1,375 @@
+import re
+from pathlib import Path
+
+from .generatorbase import GeneratorBase
+from .model import (
+    ReferenceManual,
+    Function,
+    Object,
+    Type,
+    PosArg,
+    VarArgs,
+    Kwarg,
+)
+
+import typing as T
+
+
+class ManPage:
+    def __init__(self, path: Path):
+        self.path = path
+        self.text = ""
+
+    def reset_font(self) -> None:
+        self.text += ".P\n"
+
+    def title(self, name: str, section: int) -> None:
+        import datetime
+
+        date = datetime.date.today()
+        self.reset_font()
+        self.text += f'.TH "{name}" "{section}" "{date}"\n'
+
+    def section(self, name: str) -> None:
+        self.reset_font()
+        self.text += f".SH {name}\n"
+
+    def subsection(self, name: str) -> None:
+        self.reset_font()
+        self.text += f".SS {name}\n"
+
+    def par(self, text: str) -> None:
+        self.reset_font()
+        self.text += f"{text}\n"
+
+    def indent(self) -> None:
+        self.text += ".RS 4\n"
+
+    def unindent(self) -> None:
+        self.text += ".RE\n"
+
+    def br(self) -> None:
+        self.text += ".br\n"
+
+    def nl(self) -> None:
+        self.text += "\n"
+
+    def line(self, text: str) -> None:
+        if text and text[0] in [".", "'"]:
+            self.text += "\\"
+        self.text += f"{text}\n"
+
+    def inline(self, text: str) -> None:
+        self.text += f"{text}"
+
+    def write(self) -> None:
+        self.path.write_text(self.text, encoding="utf-8")
+
+    @staticmethod
+    def bold(text: str) -> str:
+        return f"\\fB{text}\\fR"
+
+    @staticmethod
+    def italic(text: str) -> str:
+        return f"\\fI{text}\\fR"
+
+
+class GeneratorMan(GeneratorBase):
+    def __init__(
+        self, manual: ReferenceManual, out: Path, enable_modules: bool
+    ) -> None:
+        super().__init__(manual)
+        self.out = out
+        self.enable_modules = enable_modules
+        self.links: T.List[str] = []
+
+    def _extract_meson_version(self) -> str:
+        # Hack around python relative imports to get to the Meson version
+        import sys
+
+        sys.path.append(Path(__file__).resolve().parents[2].as_posix())
+        from mesonbuild.coredata import version
```

Comment:
In setuptools we handle this via:

```ini
[metadata]
version = attr: mesonbuild.coredata.version
```

Which (relies on the import system to find a filename and) reads the contents of a given file to extract an attribute without actually importing. See `from setuptools.config import StaticModule`.

Extracting the actual important bits from their importlib entanglement, you can get the version like this:

```python
import ast

with open(Path(__file__).resolve().parents[2] / 'mesonbuild/coredata.py', 'r') as f:
    src = f.read()

astdata = ast.parse(src)
version = next(
    ast.literal_eval(statement.value)
    for statement in astdata.body
    if isinstance(statement, ast.Assign)
    for target in statement.targets
    if isinstance(target, ast.Name) and target.id == 'version'
)
```

Reply:
I shamelessly copied that function from `generatorjson.py` :). https://github.com/mesonbuild/meson/blob/09f52a3817126bc1dc9d1b9d93d531505801a59e/docs/refman/generatorjson.py#L95

I wonder if there was any particular reason @mensinda decided to write it that way.  Your version seems just a little bit hacky as well. In any case, the duplication is probably bad and should be consolidated!

Reply:
This hack is an elegant methodology for parsing and iterating over an AST. :p 

No need to mess with sys.path, avoid assuming that anything is importable (although it should be) -- it's generally a nice clean way to avoid runtime requirements to retrieve version info, which is why setuptools implements it.

Yes, it should probably be moved to a more accessible location.

Reply:
So what should be done for this PR?  Should I implement your AST method in `generatorman.py` and be done with it?  I'm not really sure where an accessible version getter function should go, and it probably wouldn't be in scope for this PR anyway.

Reply:
You could put it as a staticmethod in generatorbase.py I guess.

Actually I'm a bit confused by the fact that it already imports mesonbuild.mlog a bunch of times, apparently from a previous system install of meson.

Reply:
For the `mlog` import, there is another `path` hack [here](https://github.com/mesonbuild/meson/blob/master/docs/genrefman.py#L18-L21). So no existing meson install should be required. As for the `_extract_meson_version` the additional path hack is completely unnecessary and just importing `mesonbuild.coredata` works. Seems like I forgot about the initial `sys.path` hack when I wrote the JSON generator...

Since everything from `mesonbuild` is importable already, I don't see a reason for using the `ast` workaround, when we can just import the version. In general, this hack is just a workaround to tell python: Sorry, I didn't put the root script in the git root directory because it doesn't belong there, but trust me this is still our project code and this import is fine.

To clarify, I was using `mlog` so that I could use the existing Meson printing helpers (color, nested, etc.) without duplicating any code or introducing another external dependency just for that.

Reply:
So basically, this should IMHO be fine in `GeneratorBase`, since `mlog` is already being used.

```python
    @staticmethod
    def _extract_meson_version() -> str:
        from mesonbuild.coredata import version
        return version
```

@eli-schwartz If you really don't like the `sys.path` hack, then we can try to find another solution.

Reply:
Ah, I only looked in the refman/ directory...

Using mlog is sufficient rationale for importing mesonbuild for real. The sys.path hack is fine. As such, we might as well use it for the version too.

If we were *only* doing it for the version, I'd probably opt for the ast hack instead.